Optimize PersianCalendar.AddMonths calculations

ebraminio opened the pull request:
This change add a copy of GetDatePart in PersianCalendar
just for the use of AddMonths, similar approach to DateTime::GetDate,
just to reduce PersianNewYearOnOrBefore expensive calls.

(local build of this change is in progress on my machine, pardon me beforehand if it doesn't build, I will improve accordingly due to compiler hints and yours)

tarekgh commented:
I believe the right fix here is to follow the pattern we established in the file https://github.com/dotnet/runtime/blob/3159d62aeb00e2f029f0c50e0a67b76490c408cc/src/libraries/System.Private.CoreLib/src/System/DateTime.cs#L852. So, you need to revert the changes in the GetDateParts method and have another method like [GetDate](https://github.com/dotnet/runtime/blob/3159d62aeb00e2f029f0c50e0a67b76490c408cc/src/libraries/System.Private.CoreLib/src/System/DateTime.cs#L900) which will return all parts. 

Also, I am seeing the same issue with other calendars, it will be good if we have one consistent fix for all these calendars. 

https://github.com/dotnet/runtime/blob/3159d62aeb00e2f029f0c50e0a67b76490c408cc/src/libraries/System.Private.CoreLib/src/System/Globalization/HebrewCalendar.cs#L532
https://github.com/dotnet/runtime/blob/3159d62aeb00e2f029f0c50e0a67b76490c408cc/src/libraries/System.Private.CoreLib/src/System/Globalization/HijriCalendar.cs#L278
https://github.com/dotnet/runtime/blob/3159d62aeb00e2f029f0c50e0a67b76490c408cc/src/libraries/System.Private.CoreLib/src/System/Globalization/JulianCalendar.cs#L200

ebraminio committed:
Optimize PersianCalendar.AddMonths calculations

This change add a copy of GetDatePart in PersianCalendar
just for the use of AddMonths, similar approach to DateTime::GetDate,
just to reduce PersianNewYearOnOrBefore expensive calls.

tarekgh reviewed (approved):
Thanks @ebraminio. LGTM.
